polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -7,166 +7,367 @@
+  assert(Index < ParmAttrs.size() && "Index out of range");
+  return ParmAttrs[Index];
+}
+
+//===----------------------------------------------------------------------===//
+// AttrBuilder Method Implementations
+//===----------------------------------------------------------------------===//
+
+AttrBuilder &AttrBuilder::addAttribute(llvm::Attribute::AttrKind Kind,
+                                       Optional<StringLiteral> Dialect,
+                                       AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr = createNamedAttr(
+      createStringAttr(AttrName, Dialect, Ctx), Builder.getUnitAttr());
+  return Invoke(AddAttrPtr, NamedAttr);
+}
+
+AttrBuilder &AttrBuilder::addAttribute(llvm::Attribute::AttrKind Kind,
+                                       mlir::Type Ty,
+                                       Optional<StringLiteral> Dialect,
+                                       AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr = createNamedAttr(
+      createStringAttr(AttrName, Dialect, Ctx), mlir::TypeAttr::get(Ty));
+
+  return Invoke(AddAttrPtr, NamedAttr);
+}
+
+AttrBuilder &AttrBuilder::addAttribute(llvm::Attribute::AttrKind Kind,
+                                       uint64_t Val,
+                                       AddRawIntAttrFuncPtr AddRawIntAttrPtr) {
+  assert(AddRawIntAttrPtr &&
+         "'AddRawIntAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddRawIntAttrFuncPtr AddRawIntAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddRawIntAttrPtr)(Args...);
+  };
+
+  switch (Kind) {
+  case llvm::Attribute::AttrKind::Alignment:
+  case llvm::Attribute::AttrKind::StackAlignment:
+    assert(Val <= llvm::Value::MaximumAlignment && "Alignment too large");
+    return (!Val) ? *this : Invoke(AddRawIntAttrPtr, Kind, Val);
+
+  case llvm::Attribute::AttrKind::Dereferenceable:
+  case llvm::Attribute::AttrKind::DereferenceableOrNull:
+  case llvm::Attribute::AttrKind::UWTable:
+    return (!Val) ? *this : Invoke(AddRawIntAttrPtr, Kind, Val);
```

Comment:
What's the difference between line 131 and line 136?
```suggestion
  case llvm::Attribute::AttrKind::Alignment:
  case llvm::Attribute::AttrKind::StackAlignment:
    assert(Val <= llvm::Value::MaximumAlignment && "Alignment too large");
    LLVM_FALLTHROUGH;
  case llvm::Attribute::AttrKind::Dereferenceable:
  case llvm::Attribute::AttrKind::DereferenceableOrNull:
  case llvm::Attribute::AttrKind::UWTable:
    return (!Val) ? *this : Invoke(AddRawIntAttrPtr, Kind, Val);
```